packages/rest/src/middleware/middleware-interceptor.ts

```diff
@@ -0,0 +1,221 @@
+// Copyright IBM Corp. 2020. All Rights Reserved.
+// Node module: @loopback/rest
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import {
+  asGlobalInterceptor,
+  BindingScope,
+  config,
+  Constructor,
+  Context,
+  ContextTags,
+  createBindingFromClass,
+  GenericInterceptor,
+  Interceptor,
+  InvocationContext,
+  Provider,
+} from '@loopback/core';
+import assert from 'assert';
+import debugFactory from 'debug';
+import pEvent from 'p-event';
+import {RestBindings} from '../keys';
+import {RequestContext} from '../request-context';
+import {ExpressRequestHandler} from '../router';
+import {MiddlewareFactory, MiddlewareInterceptorBindingOptions} from './types';
+
+const debug = debugFactory('loopback:rest:middleware');
+
+/**
+ * Create an interceptor function from express middleware
+ * @param middleware - Express middleware module name or factory function
+ * @param middlewareConfig - Configuration for the middleware
+ *
+ * @typeParam CFG - Configuration type
+ * @typeParam CTX - Context type
+ */
+export function createInterceptor<CFG, CTX extends Context = InvocationContext>(
+  middleware: string | MiddlewareFactory<CFG>,
+  middlewareConfig?: CFG,
+): GenericInterceptor<CTX> {
+  let factory: MiddlewareFactory<CFG>;
+  if (typeof middleware === 'string') {
+    factory = require(middleware);
```

Comment:
Please note this `require` call will resolve the path relatively to the location of `@loopback/rest` module, not relatively to the code calling `createInterceptor`. This creates confusing behavior when `createInterceptor` is called e.g. from a 3rd-party package that's symlinked into `node_modules` tree.

We have been bitten by this problem multiple times in the past. While there are ways how to work around it, I feel the complexity is not worth the benefits.

Can you please remove the string-based variant and always require the factory function instead?

Also what if the middleware does not provide any factory function and all the users have is a regular function `(req, res, next) => void`? Are they supposed to wrap that function into a factory just for the sake of satisfying this API?

Last but not least, how do you envision caching the instances of middleware between requests? It's important to call middleware factory function only once at app startup, we must not call it for every incoming request. I don't yet fully comprehend how is `createInterceptor` going to be used, so maybe this is not a problem? IDK.

I also feel this function is violating Single Responsibility Principle, because it has two responsibilities:
- Obtain an instance of middleware handler.
- Convert a middleware handler to an interceptor.

Have you considering removing the first responsibility and modifying the function to accept the middleware handler only?

Reply:
- Support for middleware module as `string` is removed.
- Default `InvokeMiddleware` is set to no-op.
- Caching is handled at binding scope level.